diff/autofd/autofd.go

```diff
@@ -0,0 +1,298 @@
+			packages.NeedFiles |
+			packages.NeedCompiledGoFiles |
+			packages.NeedSyntax |
+			packages.NeedTypes |
+			packages.NeedTypesInfo,
+	}
+	pkgs, err := packages.Load(cfg, path)
+	if err != nil {
+		return nil, fmt.Errorf("could not load package of %q: %w", f, err)
+	}
+
+	pkg := func() *packages.Package {
+		for _, p := range pkgs {
+			if p.PkgPath == path {
+				return p
+			}
+		}
+		return nil
+	}()
+
+	if pkg == nil || len(pkg.Errors) > 0 {
+		return nil, fmt.Errorf("could not find package %q", path)
+	}
+
+	var (
+		fct   *types.Func
+		scope = pkg.Types.Scope()
+	)
+	switch {
+	case strings.Contains(name, "."):
+		idx := strings.Index(name, ".")
+		obj := scope.Lookup(name[:idx])
+		if obj == nil {
+			return nil, fmt.Errorf("could not lookup %q in package %q", name[:idx], path)
+		}
+		typ, ok := obj.Type().(*types.Named)
+		if !ok {
+			return nil, fmt.Errorf(
+				"object %q in package %q is not a named type (%T)",
+				name[:idx], path, obj,
+			)
+		}
+		fct = func() *types.Func {
+			for i := 0; i < typ.NumMethods(); i++ {
+				m := typ.Method(i)
+				if m.Name() == name[idx+1:] {
+					return m
+				}
+			}
+			return nil
+		}()
+		if fct == nil {
+			return nil, fmt.Errorf("could not find %q in package %q", name, path)
+		}
+
+	default:
+		obj := scope.Lookup(name)
+		if obj == nil {
+			return nil, fmt.Errorf("could not lookup %q in package %q", name, path)
+		}
+		var ok bool
+		fct, ok = obj.(*types.Func)
+		if !ok {
+			return nil, fmt.Errorf("object %q in package %q is not a func (%T)", name, path, obj)
+		}
+	}
+
+	if !types.Identical(fct.Type(), f1x.Type()) {
+		return nil, fmt.Errorf("invalid function signature for %s", name)
+	}
+
+	der := f.Deriv
+	if der == "" {
+		der = "Deriv" + strings.Replace(f.Name, ".", "_", -1)
+	}
+
+	return &generator{w: w, pkg: pkg, fct: fct, der: der}, nil
+}
+
+func (g *generator) generate() error {
+	fct := func() *ast.FuncDecl {
+		for _, f := range g.pkg.Syntax {
+			for i := range f.Decls {
+				decl, ok := f.Decls[i].(*ast.FuncDecl)
+				if !ok {
+					continue
+				}
+
+				if decl.Name.Name == g.fct.Name() {
+					return decl
+				}
+			}
+		}
+		return nil
+	}()
+
+	var (
+		ret     *ast.ReturnStmt
+		returns = 0
+	)
+	ast.Inspect(fct.Body, func(n ast.Node) bool {
+		switch stmt := n.(type) {
+		case *ast.ReturnStmt:
+			returns++
+			ret = stmt
+		}
+		return true
+	})
+
+	if returns > 1 {
+		return fmt.Errorf("can not handle functions with multiple return statements")
+	}
+	if returns == 0 {
+		return fmt.Errorf("could not find a return statement")
+	}
+	switch len(ret.Results) {
+	case 0:
+		return fmt.Errorf("naked returns not supported")
+	case 1:
+		// ok
+	default:
+		return fmt.Errorf("too many return values")
+	}
+
+	args := g.fct.Type().Underlying().(*types.Signature).Params()
+	g.printf("func %s(%s float64) float64 {\n",
+		g.der,
+		args.At(0).Name(),
+	)
+	g.printf("\treturn ")
+	g.expr(ret.Results[0])
+	g.printf(".Emag\n")
+	g.printf("}\n")
```

Comment:
Can we make this do:
```
d := <dual.Func>()
return d.Emag
```
This would allow people to look at the result when debugging a little more easily if they need to.